File: Samples/Digital signatures/SignWithAzureKeyVault/VB.NET/AzureSigner.vb
Imports Azure.Security.KeyVault.Keys.Cryptography
Imports BitMiracle.Docotic.Pdf
Imports AzureSignatureAlgorithm = Azure.Security.KeyVault.Keys.Cryptography.SignatureAlgorithm

Class AzureSigner
    Implements IPdfSigner

    Private ReadOnly m_client As CryptographyClient
    Private ReadOnly m_signingAlgorithm As AzureSignatureAlgorithm

    Public Sub New(client As CryptographyClient, signingAlgorithm As AzureSignatureAlgorithm)
        m_client = client
        m_signingAlgorithm = signingAlgorithm
    End Sub

    Public ReadOnly Property SignatureAlgorithm As PdfSignatureAlgorithm Implements IPdfSigner.SignatureAlgorithm
        Get
            If m_signingAlgorithm = AzureSignatureAlgorithm.ES256 OrElse
                m_signingAlgorithm = AzureSignatureAlgorithm.ES384 OrElse
                m_signingAlgorithm = AzureSignatureAlgorithm.ES512 OrElse
                m_signingAlgorithm = AzureSignatureAlgorithm.ES256K Then
                Return PdfSignatureAlgorithm.Ecdsa
            End If

            If m_signingAlgorithm = AzureSignatureAlgorithm.PS256 OrElse
                m_signingAlgorithm = AzureSignatureAlgorithm.PS384 OrElse
                m_signingAlgorithm = AzureSignatureAlgorithm.PS512 Then
                Return PdfSignatureAlgorithm.RsaSsaPss
            End If

            If m_signingAlgorithm = AzureSignatureAlgorithm.RS256 OrElse
                m_signingAlgorithm = AzureSignatureAlgorithm.RS384 OrElse
                m_signingAlgorithm = AzureSignatureAlgorithm.RS512 Then
                Return PdfSignatureAlgorithm.Rsa
            End If

            Throw New InvalidOperationException($"Unsupported {NameOf(SignatureAlgorithm)} value: {m_signingAlgorithm}")
        End Get
    End Property

    Public ReadOnly Property DigestAlgorithm As PdfDigestAlgorithm
        Get
            If m_signingAlgorithm = AzureSignatureAlgorithm.RS256 OrElse
                m_signingAlgorithm = AzureSignatureAlgorithm.PS256 OrElse
                m_signingAlgorithm = AzureSignatureAlgorithm.ES256 OrElse
                m_signingAlgorithm = AzureSignatureAlgorithm.ES256K Then
                Return PdfDigestAlgorithm.Sha256
            End If

            If m_signingAlgorithm = AzureSignatureAlgorithm.RS384 OrElse
                m_signingAlgorithm = AzureSignatureAlgorithm.PS384 OrElse
                m_signingAlgorithm = AzureSignatureAlgorithm.ES384 Then
                Return PdfDigestAlgorithm.Sha384
            End If

            If m_signingAlgorithm = AzureSignatureAlgorithm.RS512 OrElse
                m_signingAlgorithm = AzureSignatureAlgorithm.PS512 OrElse
                m_signingAlgorithm = AzureSignatureAlgorithm.ES512 Then
                Return PdfDigestAlgorithm.Sha512
            End If

            Throw New InvalidOperationException($"Unsupported {NameOf(SignatureAlgorithm)} value: {m_signingAlgorithm}")
        End Get
    End Property

    Public Function Sign(message As Byte()) As Byte() Implements IPdfSigner.Sign
        Return m_client.SignData(m_signingAlgorithm, message).Signature
    End Function
End Class

File: Samples/Digital signatures/SignWithAzureKeyVault/C#/AzureSigner.cs
using System;
using Azure.Security.KeyVault.Keys.Cryptography;
using AzureSignatureAlgorithm = Azure.Security.KeyVault.Keys.Cryptography.SignatureAlgorithm;

namespace BitMiracle.Docotic.Pdf.Samples
{
    class AzureSigner : IPdfSigner
    {
        private readonly CryptographyClient m_client;
        private readonly AzureSignatureAlgorithm m_signingAlgorithm;

        public AzureSigner(CryptographyClient client, AzureSignatureAlgorithm signingAlgorithm)
        {
            m_client = client;
            m_signingAlgorithm = signingAlgorithm;
        }

        public PdfSignatureAlgorithm SignatureAlgorithm
        {
            get
            {
                if (m_signingAlgorithm == AzureSignatureAlgorithm.ES256 ||
                    m_signingAlgorithm == AzureSignatureAlgorithm.ES384 ||
                    m_signingAlgorithm == AzureSignatureAlgorithm.ES512 ||
                    m_signingAlgorithm == AzureSignatureAlgorithm.ES256K)
                    return PdfSignatureAlgorithm.Ecdsa;

                if (m_signingAlgorithm == AzureSignatureAlgorithm.PS256 ||
                    m_signingAlgorithm == AzureSignatureAlgorithm.PS384 ||
                    m_signingAlgorithm == AzureSignatureAlgorithm.PS512)
                    return PdfSignatureAlgorithm.RsaSsaPss;

                if (m_signingAlgorithm == AzureSignatureAlgorithm.RS256 ||
                    m_signingAlgorithm == AzureSignatureAlgorithm.RS384 ||
                    m_signingAlgorithm == AzureSignatureAlgorithm.RS512)
                    return PdfSignatureAlgorithm.Rsa;

                throw new InvalidOperationException($"Unsupported {nameof(SignatureAlgorithm)} value: {m_signingAlgorithm}");
            }
        }

        public PdfDigestAlgorithm DigestAlgorithm
        {
            get
            {
                if (m_signingAlgorithm == AzureSignatureAlgorithm.RS256 ||
                    m_signingAlgorithm == AzureSignatureAlgorithm.PS256 ||
                    m_signingAlgorithm == AzureSignatureAlgorithm.ES256 ||
                    m_signingAlgorithm == AzureSignatureAlgorithm.ES256K)
                    return PdfDigestAlgorithm.Sha256;

                if (m_signingAlgorithm == AzureSignatureAlgorithm.RS384 ||
                    m_signingAlgorithm == AzureSignatureAlgorithm.PS384 ||
                    m_signingAlgorithm == AzureSignatureAlgorithm.ES384)
                    return PdfDigestAlgorithm.Sha384;

                if (m_signingAlgorithm == AzureSignatureAlgorithm.RS512 ||
                    m_signingAlgorithm == AzureSignatureAlgorithm.PS512 ||
                    m_signingAlgorithm == AzureSignatureAlgorithm.ES512)
                    return PdfDigestAlgorithm.Sha512;

                throw new InvalidOperationException($"Unsupported {nameof(SignatureAlgorithm)} value: {m_signingAlgorithm}");
            }
        }

        public byte[] Sign(byte[] message)
        {
            return m_client.SignData(m_signingAlgorithm, message).Signature;
        }
    }
}
